### PR TITLE
Fix ParserCore#unexpected

### PR DESCRIPTION
**Pull request by ooflorent**

### CODE REVIEW

File: packages/@romejs/parser-core/index.ts [hunk 1]
```diff
@@ -525,12 +525,12 @@ export class ParserCore<Tokens extends TokensShape, State> {
   }
 
   // Return an error to indicate a parser error, this must be thrown at the callsite for refinement
-  unexpected(opts: ParserUnexpectedOptions = {}) {
-    throw createSingleDiagnosticError(this.createDiagnostic(opts));
+  unexpected(opts: ParserUnexpectedOptions = {}): DiagnosticsError {
+    return createSingleDiagnosticError(this.createDiagnostic(opts));
```
Comment (ooflorent) [inline packages/@romejs/parser-core/index.ts]: According to its usage, this method should return an error, not throwing it.

**Comment by sebmck** [comment]
Could you describe a bit more about the motivations for the PR? What is being fixed?

**Comment by ooflorent** [comment]
Before this PR, `unexpected` was throwing the error but the parser implementations are calling it to get a diagnostic error:

```js
// packages/@romejs/codec-json/parse.ts#L187
this.input[get0(endIndex) - 1] !== '/') {
  throw this.unexpected({
    description: descriptions.JSON.UNCLOSED_BLOCK_COMMENT,
    start: this.getPositionFromIndex(endIndex),
  });
}
```

```js
// packages/@romejs/parser-core/index.ts#L381
if (nextToken === undefined) {
  throw this.unexpected({
    start: this.getPositionFromIndex(index),
  });
}
```

In order to be consistent about its usage, I returned the error and added types to be sure the implementations are using `unexpected` correctly.

**Comment by lodenrogue** [comment]
Any tests for these changes?

**Pull request merged**